src/Raven.Server/Storage/Schema/Updates/Documents/60000/From50002.cs

```diff
@@ -0,0 +1,104 @@
+using static Raven.Server.Documents.Schemas.Documents;
+using static Raven.Server.Documents.Schemas.Revisions;
+using static Raven.Server.Documents.Schemas.TimeSeries;
+using static Raven.Server.Documents.Schemas.Tombstones;
+
+namespace Raven.Server.Storage.Schema.Updates.Documents
+{
+    public class From50002 : ISchemaUpdate
+    {
+        public int From => 50_002;
+
+        public int To => 60_000;
+        
+        public SchemaUpgrader.StorageType StorageType => SchemaUpgrader.StorageType.Documents;
+
+        internal static int NumberOfItemsToMigrateInSingleTransaction = 10_000;
+        internal static int MaxSizeToMigrateInSingleTransaction = 64 * 1024 * 1024;
+
+        public bool Update(UpdateStep step)
+        {
+            InsertIndexValuesFor(step, DocsSchemaBase60, DocsBucketAndEtagSlice);
+            InsertIndexValuesFor(step, TombstonesSchemaBase60, TombstonesBucketAndEtagSlice);
+            InsertIndexValuesFor(step, RevisionsSchemaBase60, RevisionsBucketAndEtagSlice);
+            InsertIndexValuesFor(step, TimeSeriesSchemaBase60, TimeSeriesBucketAndEtagSlice);
+            InsertIndexValuesFor(step, DeleteRangesSchemaBase60, DeletedRangesBucketAndEtagSlice);
+            InsertIndexValuesFor(step, CountersSchemaBase60, CountersBucketAndEtagSlice);
+
+            InsertIndexValuesFor(step, ConflictsSchemaBase60, ConflictsBucketAndEtagSlice,
+                fixedSizeIndex: ConflictsSchemaBase60.FixedSizeIndexes[AllConflictedDocsEtagsSlice]);
+
+            InsertIndexValuesFor(step, AttachmentsSchemaBase60, AttachmentsBucketAndEtagSlice,
+                tableName: AttachmentsMetadataSlice.ToString(),
+                fixedSizeIndex: AttachmentsSchemaBase60.FixedSizeIndexes[AttachmentsEtagSlice]);
+
+            return true;
+        }
+
+        public static void InsertIndexValuesFor(UpdateStep step, TableSchema schema, Slice indexName, string tableName = null, TableSchema.FixedSizeKeyIndexDef fixedSizeIndex = null)
+        {
+            var indexTree = step.WriteTx.CreateTree(indexName, isIndexTree: true);
+            var indexDef = schema.DynamicKeyIndexes[indexName];
+            
+            bool done = false;
+            long skip = 0;
+
+            while (done == false)
+            {
+                var table = tableName == null
+                    ? new Table(schema, step.ReadTx)
+                    : step.ReadTx.OpenTable(schema, tableName);
+
+                var processedInCurrentTx = 0;
+
+                using (step.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                {
+                    context.TransactionMarkerOffset = (short)step.WriteTx.LowLevelTransaction.Id;
+
+                    var commit = false;
+
+                    var items = fixedSizeIndex == null
+                        ? table.SeekByPrimaryKey(Slices.BeforeAllKeys, skip)
+                        : table.SeekForwardFrom(fixedSizeIndex, 0, skip);
```

Comment:
We're always doing the seek to the first item and then skipping. The implementation of the skip operation might be actually pretty expensive when we're skipping to the records at the end because we'll iterate over the whole tree. Please check the skip impl:

https://github.com/ravendb/ravendb/blob/3d10ece8670a53f350b10db7460240f27219a345/src/Voron/Data/Fixed/FixedSizeIterators.cs#L317-L372

We should be able to Seek to close to the last record where we ended (optionally skip few item if needed to be on a good position) and then continue the process.

Reply:
In general Skip is supposed to deal with relatively small number of items. Not to skip millions of records.

Reply:
So we simply need to keep the last processed key?

Reply:
fixed